build/accelerate_launch.py

```diff
@@ -0,0 +1,94 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Script wraps launch_training to run with accelerate for multi and single GPU cases.
+Read multiGPU configuration via environment variable `SFT_TRAINER_CONFIG_JSON_PATH`
+for the path to the JSON config file with parameters or `SFT_TRAINER_CONFIG_JSON_ENV_VAR`
+for the encoded config string to parse.
+"""
+
+# Standard
+import json
+import os
+import base64
+import pickle
+import logging
+
+# Third Party
+from accelerate.commands.launch import launch_command_parser, launch_command
+import torch
+
+
+def txt_to_obj(txt):
+    base64_bytes = txt.encode("ascii")
+    message_bytes = base64.b64decode(base64_bytes)
+    try:
+        # If the bytes represent JSON string
+        return json.loads(message_bytes)
+    except UnicodeDecodeError:
+        # Otherwise the bytes are a pickled python dictionary
+        return pickle.loads(message_bytes)
+
+
+def main():
+    LOGLEVEL = os.environ.get("LOG_LEVEL", "WARNING").upper()
+    logging.basicConfig(level=LOGLEVEL)
+
+    json_configs = {}
+    json_path = os.getenv("SFT_TRAINER_CONFIG_JSON_PATH")
+    json_env_var = os.getenv("SFT_TRAINER_CONFIG_JSON_ENV_VAR")
+
+    if json_path:
+        with open(json_path, "r", encoding="utf-8") as f:
+            json_configs = json.load(f)
+
+    elif json_env_var:
+        json_configs = txt_to_obj(json_env_var)
+
+    # parse multiGPU args
+    multi_gpu_args = []
+    if json_configs.get("multiGPU"):
+        logging.info("Using multi-GPU configs: %s", json_configs.get("multiGPU"))
+        for key, val in json_configs["multiGPU"].items():
+            multi_gpu_args.append(f"--{key}")
+            multi_gpu_args.append(str(val))
+
+        # add FSDP config
+        if not json_configs.get("multiGPU").get("config_file"):
+            fsdp_filepath = os.getenv(
+                "FSDP_DEFAULTS_FILE_PATH", "/app/accelerate_fsdp_defaults.yaml"
+            )
+            if os.path.exists(fsdp_filepath):
+                logging.info(f"Setting accelerate config file to: {fsdp_filepath}")
+                multi_gpu_args.append("--config_file")
+                multi_gpu_args.append(fsdp_filepath)
+
+        # add num_processes to overwrite config file set one
+        if not json_configs.get("multiGPU").get("num_processes"):
+            num_gpus = torch.cuda.device_count()
+            if num_gpus > 1:
+                logging.info(f"Setting accelerate num processes to {num_gpus}")
+                multi_gpu_args.append("--num_processes")
+                multi_gpu_args.append(str(num_gpus))
```

Comment:
Note that here I have set that we are using `accelerate_fsdp_defaults.yaml` as the default config file, and this file requires setting `num_processes`. So we have to overwrite this value which is set to 1 by default by sending a flag with the value.

If we didn't have to set the FSDP config file, then the accelerate script knows to calculate num of GPUs available and to set that as the num_processeses. accelerate launch script does this calculation [here](https://github.com/huggingface/accelerate/blob/main/src/accelerate/commands/launch.py#L974-L988) and I simplified this to just use `torch.cuda.device_count()`

Reply:
what if `json_configs.get("multiGPU")` returns `None`?

Reply:
@anhuong btw if you are linking to HF you should use a permalink, because HF merges everyday.

Reply:
@fabianlim In the bullets do you mean `args.num_processes` since I am not touching the `args.num_machines` in the script? 

`accelerate launch` is a smart script and does a lot of processing of the params passed in. If `num_processes` is never specified, then it can determine how many resources are available and assign `num_processes` appropriately. However, since the `config_file` specifies `num_processes` to be 1,  I am overriding that value to the number of GPUs available if multiGPU is set. If multi-GPU is not set, I explicitly set `num_processes` to be 1 in case the pod has more than 1 GPU allocated, in which if `num_processes` is not passed in as a flag, `accelerate launch` would assume multiGPU if multiple GPUs are available given that no config_file is passed and no flag is passed but multiple GPUs exist.

. This can be overwritten if a flag is passed in

Reply:
Ah I see this comment is further discussed below and seems resolved. Let me know if you have more questions about it